lib/pure/options.nim

```diff
@@ -176,21 +310,239 @@ proc filter*[T](self: Option[T], callback: proc (input: T): bool): Option[T] =
+          "previously defined \"none\" case at: " & lineInfo(noneCase),
+          optionCase[0]
+      else:
+        noneCase = optionCase[1]
+    of nnkCommand:
+      if $optionCase[0] != "some":
+        if $optionCase[0] != "none":
+          error "Only \"none\" and \"some\" are allowed as case labels",
+            optionCase[0]
+        else:
+          error "Only \"some\" is allowed to have arguments", optionCase[0]
+      else:
+        if optionCase[1].kind != nnkInfix:
+          if elseCase.body != nil:
+            error "Only one \"some\" case without a check is allowed, " &
+              "previously defined \"some\" case at: " & lineInfo(elseCase.body),
+              optionCase[0]
+          else:
+            if $optionCase[1] != "_":
+              elseCase = (ident: optionCase[1], body: optionCase[2])
+            else:
+              elseCase = (ident: nil, body: optionCase[2])
+        else:
+          if $optionCase[1][0] != "of":
+            error "Invalid separator, only \"of\" allowed", optionCase[1][0]
+          else:
+            if $optionCase[1][1] != "_":
+              cases.add (ident: optionCase[1][1], check: optionCase[1][2],
+                body: optionCase[2])
+            else:
+              cases.add (ident: nil, check: optionCase[1][2],
+                body: optionCase[2])
+    else:
+      error "Unrecognized structure of cases", optionCase
+  if noneCase == nil:
+    error "Must have a \"none\" case"
+  if elseCase.body == nil and cases.len == 0:
+    error "Must have a \"some\" case"
+  var
+    optionSym = genSym(nskLet)
+    value = nnkDotExpr.newTree(optionSym, newIdentNode("val"))
+  var caseStmt = nnkCaseStmt.newTree(
+    value
+  )
+  for c in cases:
+    var ofBranch = nnkOfBranch.newTree(
+      c.check,
+      newStmtList()
+    )
+    if c.ident != nil:
+      ofBranch[1].add newLetStmt(c.ident, value)
+    ofBranch[1].add c.body
+    caseStmt.add ofBranch
+  if elseCase.body != nil:
+    if elseCase.ident != nil:
+      caseStmt.add nnkElse.newTree(newStmtList(
+        newLetStmt(elseCase.ident, value),
+        elseCase.body
+      ))
+    else:
+      caseStmt.add nnkElse.newTree(elseCase.body)
+  result = quote do:
+    (proc (): auto =
+      let `optionSym` = `option`
+      if `optionSym`.isSome:
+        `caseStmt`
+      else:
+        `noneCase`
+    )()
+  #echo result.repr
+
+proc optCmp*[T](self: Option[T], value: Option[T],
+  cmp: proc (val1, val2: T): bool): Option[T] =
+  ## Comparison operation that return a some iff ``self`` is a some and the
+  ## value of ``cmp`` applied to ``self`` and ``value`` returns true. The
+  ## returned some has the value of ``self``.
+  if self.isSome and value.isSome and cmp(self.val, value.val):
+    return self
+
+template `==`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``==`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 == val2)
```

Comment:
this is an unnecessary lambda expression. Lambda expressions have a bad performance penalty, because they can never be inlined. Comparison is something that is very likely to benefit from inlining.

Reply:
How would this be written then? I just went with the first thing that worked..